### PR TITLE
Fix organization deletion/namespace

### PR DESCRIPTION
**Pull request by ssyno**
## Checklist
This PR:
- implements fixes on `organization/namespace` deletion

- [x] Update changelog in CHANGELOG.md.

### CODE REVIEW

File: internal/controller/organization_controller.go [hunk 1]
```diff
@@ -126,19 +127,38 @@ func (r *OrganizationReconciler) Reconcile(ctx context.Context, req ctrl.Request
 func (r *OrganizationReconciler) reconcileDelete(ctx context.Context, organization *securityv1alpha1.Organization) (ctrl.Result, error) {
 	log := log.FromContext(ctx)
 
-	originalOrg := organization.DeepCopy()
-	controllerutil.RemoveFinalizer(organization, "organization.giantswarm.io/finalizer")
-	patch := client.MergeFrom(originalOrg)
-	if err := r.Patch(ctx, organization, patch); err != nil {
-		return ctrl.Result{}, fmt.Errorf("failed to remove finalizer: %w", err)
+	namespaceName := fmt.Sprintf("org-%s", organization.Name)
```
Comment (fiunchinho) [inline internal/controller/organization_controller.go:130]: You can take the namespace name from the `status` field. That way, we don't repeat the logic of how the namespace name is built.

File: internal/controller/organization_controller.go [hunk 2]
```diff
@@ -126,19 +127,38 @@ func (r *OrganizationReconciler) Reconcile(ctx context.Context, req ctrl.Request
...
+		if err := r.Delete(ctx, namespace); err != nil {
+			log.Error(err, "Failed to delete associated namespace")
+			return ctrl.Result{Requeue: true}, err
+		}
+		log.Info("Namespace deletion triggered, requeuing")
+		return ctrl.Result{Requeue: true}, nil
+	} else if !errors.IsNotFound(err) {
+		log.Error(err, "Failed to check for associated namespace")
+		return ctrl.Result{Requeue: true}, err
+	}
```
Comment (fiunchinho) [inline internal/controller/organization_controller.go:143]: I think we could call `Delete(ctx, namespace)` without having to `Get()` first, and skip that request to the API. If `Delete()` returns `NotFound`, then we can proceed to remove the finalizer. If it doesn't return any error, we `requeue`.

By the way, when returning a non nil `err`, the event will be `requeued` automatically. There is no need to set `Requeue: true` when returning a non nil `err`.

File: internal/controller/organization_controller.go [hunk 3]
```diff
@@ -126,19 +127,38 @@ func (r *OrganizationReconciler) Reconcile(ctx context.Context, req ctrl.Request
...
+
+	if controllerutil.ContainsFinalizer(organization, "organization.giantswarm.io/finalizer") {
+		log.Info("Removing finalizer from Organization")
+		originalOrg := organization.DeepCopy()
+		controllerutil.RemoveFinalizer(organization, "organization.giantswarm.io/finalizer")
+		patch := client.MergeFrom(originalOrg)
+		if err := r.Patch(ctx, organization, patch); err != nil {
+			log.Error(err, "Failed to remove finalizer")
+			return ctrl.Result{Requeue: true}, err
```
Comment (fiunchinho) [inline internal/controller/organization_controller.go:152]:
```suggestion
			return ctrl.Result{}, err
```

File: internal/controller/organization_controller_test.go [hunk 4]
```diff
@@ -116,15 +114,29 @@ var _ = Describe("Organization controller", func() {
...
-			})
-			Expect(err).NotTo(HaveOccurred())
+			Eventually(func() error {
+				err := k8sClient.Get(ctx, client.ObjectKey{Name: "test-1"}, &securityv1alpha1.Organization{})
+				if errors.IsNotFound(err) {
+					return nil
+				}
+				if err != nil {
+					return err
+				}
```
Comment (fiunchinho) [inline internal/controller/organization_controller_test.go:124]: I think you can use this instead.

```suggestion
				Expect(err).NotTo(HaveOccurred())
```

File: internal/controller/organization_controller_test.go [hunk 5]
```diff
@@ -116,15 +114,29 @@ var _ = Describe("Organization controller", func() {
...
+				// Trigger reconciliation if the organization still exists
+				_, reconcileErr := reconciler.Reconcile(ctx, reconcile.Request{
+					NamespacedName: types.NamespacedName{Name: "test-1"},
+				})
+				if reconcileErr != nil {
+					return reconcileErr
+				}
```
Comment (fiunchinho) [inline internal/controller/organization_controller_test.go:131]:
```suggestion
				Expect(reconcileErr).NotTo(HaveOccurred())
```

File: internal/controller/organization_controller.go [hunk 6]
```diff
@@ -126,19 +127,38 @@ func (r *OrganizationReconciler) Reconcile(ctx context.Context, req ctrl.Request
...
+		}
+		log.Info("Namespace deletion triggered, requeuing")
+		return ctrl.Result{Requeue: true}, nil
+	} else if !errors.IsNotFound(err) {
+		log.Error(err, "Failed to check for associated namespace")
+		return ctrl.Result{Requeue: true}, err
+	}
+
+	if controllerutil.ContainsFinalizer(organization, "organization.giantswarm.io/finalizer") {
```
Comment (fiunchinho) [inline internal/controller/organization_controller.go:145]: Do we have a test that tests the deletion of the finalizer? If we don't, I think it makes sense to add it.

**Comment by ssyno** [comment]
pushed some changes based on your comments, seems that the PR is stuck for now, but the branch has the changes updated.

**Review by fiunchinho — APPROVED** [review]

**Pull request merged**